### ISSUE TITLE
Occluding more than one sliding window 

### ISSUE THREAD

**Issue by FL33TW00D** [pytorch/captum#387]
Is it possible to occlude more than one hyperrectangle from a single input to a network?
I would like to both the row and the column of an input one by one.

**Comment by bilalsal** [comment]
Hi @FL33TW00D.
The Occlusion algorithm in Captum derives from the more generic FeatureAblation class.
This class enables you to define ablation masks that you can use to group the pixels you want to occlude at once.
This tutorial illustrate how to define custom masks: https://captum.ai/tutorials/Resnet_TorchVision_Ablation

Hope this brings you closer to your goal

**Comment by FL33TW00D** [comment]
Hi @bilalsal,
Thank you for the direction to the FeatureAblation class.
I have attempted to use it to ablate every row and column in my matrix.
Is accumulating the outputted gradients fundamentally sound, such as in this?

``    for sample in testloader:
        for i in range(batch_size):
            func, anat, label = Variable(sample['func'][i]), Variable(sample['anat'][i]), Variable(sample['label'][i])
            input = func.unsqueeze(0)
            input = input.unsqueeze(0)
            anat = anat.unsqueeze(0)
            input.requires_grad = True
            anat.requires_grad = True
            output = model(input, anat)
            output = F.softmax(output, dim=1)
            prediction_score, pred_label_idx = torch.topk(output, 1)
            
            attributions = 0
            ablator = FeatureAblation(model)
            for i in range(249):
                feature_mask = torch.ones(249,249, dtype=torch.int32)
                feature_mask[i][:] = 0
                feature_mask[:, i] = 0
                attributions += ablator.attribute(
                    input,
                    target=pred_label_idx,
                    feature_mask=feature_mask,
                    additional_forward_args=anat)

Many Thanks

**Comment by bilalsal** [comment]
Hi @FL33TW00D,

technically your code is sound.
To which extent the computed attribution is helpful and insightful, depends on the problem you are solving and which features you expect your model to pick on.
Note that the attribution you are computing is not gradient based, but ablation based.
When ablating images using Occlusion, we generally recommend using a sliding window that is big enough to occlude object features in a way that changes the visual appearance and make the object harder to detect (even for human). Occluding one pixel usually falls short of doing this.
We recommend you consider increasing the size of the stripes you are occluding, e.g.:

```python
            for i in range(0:249:10):
                feature_mask[i:i+10][:] = 0
                feature_mask[:, i+10] = 0
```

Hope this helps

**Comment by FL33TW00D** [comment]
Hi @bilalsal ,
Thank you for your help. I am not working with images I am working with matricies with each row and column representing segments of the human brain.

Keep up the great work with Captum!

**Comment by vivekmig** [comment]
Hi @FL33TW00D , as @bilalsal mentioned, Feature Ablation doesn't use any gradient information, so requiring grads in your code shouldn't be necessary. One potential issue in your current code is that with a feature mask like this:
```
0  0  0  ....
0  1  1
0  1  1 ...
 ....

```
Feature Ablation first ablates all inputs with mask value being 0 (first row and column) by setting them equal to 0 (or the specified baseline) and computes the change in output and then repeats the process setting inputs with mask value being 1 (areas other than first row and column) to 0 and computes the change in output.

The output attributions will have the same value in each cell of the first row and column corresponding to the change when ablating that area, and the remaining cells will have a separate value corresponding to the difference when ablating the remaining area.

In your case, I think you're interested in the output change for the area with mask value being 0, but not necessarily the remaining area, which is still being added in your total attributions.

To get around this, one way is to just extract the value of attributions[i][i] in your for loop (or any index in row or column i, which would be the same), which is the output change when ablating row and column i. 
 
This call to FeatureAblation is still doing an extra computation of ablating the remaining area, which may be unnecessary, so alternatively you can do the same computation done by ablation directly (not using Captum) with something like:
`model(input)[:, pred_label_idx] - model(feature_mask * input)[:, pred_label_idx]`

The feature_mask with FeatureAblation is most useful with non-overlapping ablation areas, where you can set different indices to each region. If this is useful in your context, you could use a feature mask of something like this: 
```
0  0  0  0 ....
0  1  1  1
0  1  2  2 ...
0  1  2  3
....
```
This wouldn't ablate full rows and columns, so for that use case, the suggestions above may be helpful. Hope this helps!
